lxd/storage/drivers/remote_storage_helpers.go

```
@@ -0,0 +1,159 @@
+package drivers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/canonical/lxd/lxd/operations"
+	"github.com/canonical/lxd/lxd/storage/connectors"
+	"github.com/canonical/lxd/lxd/storage/filesystem"
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/logger"
+	"github.com/canonical/lxd/shared/revert"
+	"golang.org/x/sys/unix"
+)
+
+// This file here is to consolidate all useful helpers to reuse across different remote storage
+// drivers which use NVMe/iSCSI/etc connectors like PowerFlex, Pure, HPE Alletra.
+
+// remoteDriver is the extended internal interface for use only with helpers below.
+type remoteDriver interface {
```

Why not just adding those funcs to the [`common`](https://github.com/canonical/lxd/blob/main/lxd/storage/drivers/driver_common.go#L28) driver implementation? It's already embedded by all the storage drivers.

Maybe I am missing something but I think the drivers can then still implement their own version of `unmapVolume` and `getMappedDevPath`.

---

Mhh I see. My thinking was to just have noop implementation on the `common` struct (returning not implemented like some of the others already do) and then it's up to the driver to decide implementing them.

---

So to summarise:

- Split up the remote specific code from the generic code
- Generic code that can be duplicated can go into the `common` driver function or if not wide enough in scope into a helper function into well-named sub-package.
- Remote specific code helpers should be factored out into a well-named package (or into an existing package).
- Avoid adding generic functionality into a remote specific helper function (like `UnmountRemoteVolume` is doing).

---

>For example in UnmountRemoteVolume I don't see anything specific to "remote" pools in there, and that seems to be a good candidate for being added to the common driver.

What is specific is that these functions require a driver to implement `getMappedDevPath`/`mapVolume`/`unmapVolume` methods. Which makes absolutely no sense for a "common" driver. That's was the only reason why I decided to introduce a new interface, because it feels like a remote storage is a less abstract concept than a common driver with it's own specifics and I didn't want to bring that specifics into a common interface.

>I also see no need for a new interface, but I think this is just being done to avoid import loops, which is a sign that its likely in the wrong place too.

I hadn't seen any import loops as it's all anyways sits in the same package.

---

> Do I understand you right that when running `_, deactivate, err := devPath(vol, false)` the `deactivate` func should be a noop in case the vol is not yet mounted.
> But running the same again when the vol is mounted causes `deactivate` to contain the necessary steps to unmap the vol?

yes exactly

---

Dear friends,

thanks for all the input in there. I've completely reworked this as you suggested.

The only thing is that I kept the behavior of `getMappedDevPath(vol, false)` to return an error in case if volume is not mapped. Because, just from a function name caller expects to get a mapped device path and if device is not mapped then we should return an error, IMHO. Let me know if you disagree with this and I can fix this.